aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -168,32 +169,29 @@ Tensor& cumprod_out(Tensor& result, const Tensor& self, int64_t dim, c10::option
   return result;
 }
 
-static Tensor sum_scan_exclusive(const Tensor& x, int64_t dim) {
-  Tensor ret = at::cumsum(-x, dim);
-
-  int64_t end_idx = ret.size(dim) - 1;
-  Tensor ret_sum = ret.narrow(dim, end_idx, 1).clone(at::MemoryFormat::Preserve);
-  ret -= ret_sum.expand_as(ret);
-  ret += x;
-  return ret;
+Tensor reversed_cumsum(const Tensor& w, int64_t dim) {
+  /* Implements w.flip(dim).cumsum(dim).flip(dim) without copying.
+     This implementation gives a similar performance on GPU but a
+     noticeable (up to x10) speed-up on CPU.
+   */
+  const auto w_cumsum = w.cumsum(dim);
+  const auto w_sum = w_cumsum.narrow(dim, -1, 1);
+  return w_sum - w_cumsum + w;
 }
 
-Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim) {
+Tensor cumprod_backward(const Tensor& grad, const Tensor& input, int64_t dim, const Tensor& output) {
   /*
-    There are two algorithms to do this. The first one
-    is very efficient, but works only when there are no
-    nonzero elements in the input.
+    We show here how to derive an O(n) gradient formula for
+    abitrary inputs. It follows via a basic application of the
+    chain rule together with a number of observations for different
+    cases. We assume that x is an n-dimensional vector and y = cumprod(x).
+    In the actual implementation we will need to play a bit with masks
+    to be able to implement the formulas deduced here for tensors.
 
-    The second one is much more complex, but it doesn't
-    assume anything on the input. The main downside is
-    that it takes time O(n^2), where n = input.size(self.dim)
-    (i.e. the length of the cumulative product). This is in
-    contrast to the forward pass and the efficient algorithm,
-    which are both O(n).
+    We will first deduce the formula for the case when
+    x[i] != 0 for 1 <= i <= n.
 
-    The second algorithm is a simple application of the chain
-    rule. If x is an n-dimensional vector, and y = cumprod(x),
-    and F is the final cost, then
+    For F : R^n -> R the cost function, we have
```

Comment:
nit F: H^n -> H where F is the cost function for a real or complex field H ...

Reply:
This PR does not add complex support just yet. Also, it should be `F : K^n -> R` for `K = R or C`, as this is the cost function, right?

Reply:
yeah although since https://github.com/pytorch/pytorch/pull/48125 will be merged soon, I think it's ok to preemptively add that

Reply:
In fact, to avoid having to rewrite the derivation for the whole thing in the complex case, I will just add a note showing how to compute the complex gradient at the end.